app/javascript/lib/validation/expression-matrices-validation.js

```diff
@@ -0,0 +1,265 @@
+/**
+* @fileoverview Functions used for sparse (MTX), features, barcodes and dense file validation
+*/
+
+import {
+  getParsedHeaderLines, parseLine,
+  validateUniqueCellNamesWithinFile, validateMetadataLabelMatches, validateGroupColumnCounts
+} from './shared-validation'
+
+/** Parse a dense matrix file */
+export async function parseDenseMatrixFile(chunker, mimeType, fileOptions) {
+  const { headers, delimiter } = await getParsedHeaderLines(chunker, mimeType)
+  let issues = validateDenseHeader(headers)
+
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
+    const line = parseLine(rawLine, delimiter)
+    issues = issues.concat(validateValuesAreNumeric(line, isLastLine, lineNum, dataObj))
+    issues = issues.concat(validateColumnNumber(line, isLastLine, headers, lineNum, dataObj))
+    issues = issues.concat(validateUniqueCellNamesWithinFile(line, isLastLine, dataObj))
+    issues = issues.concat(validateMetadataLabelMatches(headers, line, isLastLine, dataObj))
+    issues = issues.concat(validateGroupColumnCounts(headers, line, isLastLine, dataObj))
+    // add other line-by-line validations here
+  })
+  return { issues, delimiter, numColumns: headers[0].length }
+}
+
+/** Parse an MTX matrix file */
+export async function parseSparseMatrixFile(chunker, mimeType, fileOptions) {
+  let issues = []
+  const delimiter = '  ' // tab delimited
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
+    issues = issues.concat(validateMTXHeaderLine(rawLine, lineNum, dataObj))
```

Comment:
Blocker: validating the header on every line is slower than necessary -- it can foreseeably be > 1 million unneeded calls each to `concat`, `validateMTXHeaderLine`, and downstream array initialization and `slice`.

Ideally, we'd use an approach that's consistent in form with that in `parseDenseMatrixFile` -- i.e. hoisting the logic for header validation up and out of the expensive `iterateLines` loop.  I see how `getParsedHeaderLines` might not work for MTX files, so I'd adjust that header-parsing code for this use case.  

Perhaps something like:

```suggestion
  let rawHeaderLine = null
  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
    rawHeaderLine = rawLine
  }, 1)
  let issues = validateMTXHeaderLine(rawHeaderLine)
  
  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
```